### PR TITLE
Fix the remaining txns calculation in proof queue

### PR DESCRIPTION
**Pull request by vusirikala**
## Description
Fix the remaining txns calculation in proof queue

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by bchocho** [comment]
@vusirikala in `realistic_env_max_load` I still see the unique being counted more than the total

File: testsuite/forge-cli/src/main.rs [hunk 1]
```diff
@@ -1952,9 +1953,9 @@ fn realistic_env_max_load_test(
         .add_system_metrics_threshold(SystemMetricsThreshold::new(
             // Check that we don't use more than 18 CPU cores for 10% of the time.
             MetricsThreshold::new(18.0, 10),
-            // Memory starts around 3GB, and grows around 1.2GB/hr in this test.
-            // Check that we don't use more than final expected memory for more than 10% of the time.
-            MetricsThreshold::new_gb(3.3 + 1.4 * (duration_secs as f64 / 3600.0), 10),
+            // Memory starts around 3.5GB, and grows around 1.5GB/hr in this test.
+            // Check that we don't use more than final expected memory for more than 20% of the time.
+            MetricsThreshold::new_gb(3.5 + 1.5 * (duration_secs as f64 / 3600.0), 20),
```
Comment (bchocho) [inline testsuite/forge-cli/src/main.rs:1958]: Is this change necessary after #13906? I'm thinking it might just be a side effect of having increased the block sizes inadvertently.

Comment (vusirikala) [reply testsuite/forge-cli/src/main.rs:1958]: You are right. If the other PR lands first, this may not be required.

**Review by bchocho — APPROVED** [review]

**Comment by sitalkedia** [comment]
@vusirikala - Can you repro the issue we were seeing on testnet and ensure that is resolved with this change?

**Comment by vusirikala** [comment]
> @vusirikala - Can you repro the issue we were seeing on testnet and ensure that is resolved with this change?

@sitalkedia The issue on the testnet wa

**Comment by vusirikala** [comment]
> @vusirikala - Can you repro the issue we were seeing on testnet and ensure that is resolved with this change?

@sitalkedia On the testnet, 2 nodes were stuck for some reason and stopped receiving commit notifications. So, their proof queue became large for those nodes. This happened even with the first tapos test on testnet before the proof queue changes were part of branch cut. With forge overload test, I made sure the formula used for "remaining unique txns" is correct and is lower than the "remaining txns" count. But if some nodes are stuck and stopped receiving commit notifications, the proof queue for those nodes will still be large.

https://aptoslabs.grafana.net/d/SnI98V7Vz/quorum-store?from=1719612962972&to=1719691645265&var-Datasource=fHo-R604z&var-metrics_source=telemetry-service&var-chain_name=testnet&var-cluster=All&var-namespace=All&var-kubernetes_pod_name=All&var-role=validator&var-interval=$__auto_interval_interval&orgId=1

File: consensus/src/quorum_store/utils.rs [hunk 2]
```diff
@@ -249,19 +253,30 @@ impl ProofQueue {
...
+            .values()
+            .map(|batches| {
+                batches
+                    .keys()
+                    .map(|batch_sort_key| {
+                        if !self.batch_summaries.contains_key(&batch_sort_key.batch_key) {
+                            if let Some(Some((proof, _))) =
+                                self.batch_to_proof.get(&batch_sort_key.batch_key)
+                            {
+                                // The batch is included in author_to_batches, batch_to_proof, but not in batch_summaries
```
Comment (sitalkedia) [inline consensus/src/quorum_store/utils.rs:273]: This will create the duplication issue - can you explain when this happens and anything we can do to resolve it?

File: consensus/src/quorum_store/utils.rs [hunk 3]
```diff
@@ -305,24 +335,59 @@ impl ProofQueue {
...
+                    (
+                        txn_summaries,
+                        batch_info.expiration(),
+                        proof_exists_in_author_to_batches && proof_exists_in_batch_to_proof,
+                    ),
+                );
+            }
         }
+
+        self.gc_expired_batch_summaries_without_proofs();
```
Comment (sitalkedia) [inline consensus/src/quorum_store/utils.rs:377]: Will this be too expensive to call for every insert? Should we be doing this periodically instead @bchocho ?

Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs:377]: Here we are iterating through the batches and checking the expiration times. We are not iterating through the transactions. This should be pretty cheap.

Comment (sitalkedia) [reply consensus/src/quorum_store/utils.rs:377]: Still it creates o(n2) operations when n is the number of batches - each insert we iterate through all batches. We should avoid o(n2) operations here.

File: consensus/src/quorum_store/utils.rs (hunk 2 above)
Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs:273]: Creating a proof needs signatures from the quorum. The other validators may not have the batch available locally. So, it's possible that a validator doesn't receive the batch summary, but receives the proof for the batch. In that case, we have to add proof.num_txns() to the calculation without any deduplication. 

If the validator receives the batch summary after receiving the proof, and then calculate the remaining txns, that's still fine. The code will do deduplication when calculating remaining txns.

File: consensus/src/quorum_store/utils.rs (hunk 3 above)
Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs:377]: I tested the code with graceful overload test. This is the latency for add batch summaries with the above garbage collection included. 
Screenshot 2024-07-09 at 12 29 13 AM

Comment (bchocho) [reply consensus/src/quorum_store/utils.rs:377]: @vusirikala is there a good reason to be doing this on every batch receive? Can we just update on the commit notification?

Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs:377]: We observed on the testnet that due to some reason, the node failed to receive commit notifications, the node got stuck and the proof queue became large. So, I did the garbage collection here instead.

Comment (bchocho) [reply consensus/src/quorum_store/utils.rs:377]: Ah, I see. Yes, this can help with nodes that are lagging on state sync.

I wonder if we can run it less frequently, maybe just log the latest time it was run and only re-run if > 200 ms or so have passed?

Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs]: Sure. I replaced the statement with `sample!` to do the garbage collection once every 500ms.

**Review by sitalkedia — APPROVED** [review]
LG
